### PR TITLE
fix: issue where subprocess provider would crash in the background

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

Fixes issue preventing foundry from working, preventing from Hardhat from working well, and currently preventing Starknet tests from passing.

Instead of sending output back to parent process, send it to a file and then everything is fine.

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape/api/providers.py [hunk 1]
```diff
@@ -921,7 +935,22 @@ def start(self, timeout: int = 20):
         else:
             logger.info(f"Starting '{self.process_name}' process.")
             pre_exec_fn = _linux_set_death_signal if platform.uname().system == "Linux" else None
-            self.process = _popen(*self.build_command(), preexec_fn=pre_exec_fn)
+
+            self.base_logs_path.mkdir(parents=True, exist_ok=True)
+            if self.stdout_logs_path.is_file():
+                self.stdout_logs_path.unlink()
+            if self.stderr_logs_path.is_file():
+                self.stderr_logs_path.unlink()
```
Comment (ca11ab1e) [inline src/ape/api/providers.py:943]:
```suggestion
            self.stdout_logs_path.unlink(missing_ok=True)
            self.stderr_logs_path.unlink(missing_ok=True)
```

File: src/ape/api/providers.py [hunk 2]
```diff
@@ -921,7 +935,22 @@ def start(self, timeout: int = 20):
...
+            self.stdout_logs_path.touch()
+            self.stderr_logs_path.touch()
+            self.process_stdout = open(str(self.stdout_logs_path))
+            self.process_stderr = open(str(self.stderr_logs_path))
```
Comment (ca11ab1e) [inline src/ape/api/providers.py:947]:
```suggestion
            self.process_stdout = self.stdout_logs_path.open()
            self.process_stderr = self.stderr_logs_path.open()
```

File: src/ape/api/providers.py (hunk 1 above)
Comment (unparalleled-js) [reply src/ape/api/providers.py:943]: > 

we can't actually do this yet because of python 3.7

Comment (ca11ab1e) [reply src/ape/api/providers.py:943]: Ah yes. That's a shame :)

**Review by johnson2427 — APPROVED** [review]
I'm approving the changes as they are. It does solve a collision issue with PIPE, but I'm not sure this solves our issue with how slow .cairo files are in ape-starknet. If we approve this, we can focus on that same slow test. I do think this will prevent that infinite loop we were seeing in ape-starknet

File: src/ape/api/providers.py [hunk 3]
```diff
@@ -860,6 +864,9 @@ class SubprocessProvider(ProviderAPI):
     process: Optional[Popen] = None
     is_stopping: bool = False
 
+    stdout_queue: Optional[JoinableQueue] = None
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:867]: At first tried naming these `_stdout_queue` etc.. but then Hardhat complains about the base classes conflictling.
Lol!

**Comment by unparalleled-js** [comment]
- [ ] hmm, what about multiple provider runnings of the same type, won't they overwrite each other's logs?

File: src/ape/api/providers.py [hunk 4]
```diff
@@ -931,6 +979,31 @@ def start(self, timeout: int = 20):
...
+            self._stderr_queue.put(line)
+            time.sleep(0)
+
+    def consume_stdout_queue(self):
+        for line in self._stdout_queue:
+            output = line.decode("utf8").strip()
+            logger.debug(output)
+            self._stdout_logger.info(output)
+            self._stdout_queue.task_done()
+            time.sleep(0)
```
[the remark this answers is not in the document]
Comment (johnson2427) [reply src/ape/api/providers.py:998]: time.sleep blocks the main thread. Could be the source of our slowing down of tests

**Comment by unparalleled-js** [comment]
> * [x]  hmm, what about multiple provider runnings of the same type, won't they overwrite each other's logs?
> 
> Edit: decided it makes more sense to handle this as part of the `x-dist` ticket

**Review by johnson2427 — APPROVED** [review]
Starknet Tests passing

**Pull request merged**